ДЗ-3:

In48semenov opened the pull request:
1. Побита метрика MAP@10: лучшая модель userKNN_bmp25 (оффлайн рекомендации с библиотекой rectools) 0.10384918
2. Сделано количество рекомендаций не меньше N: добавляли популярное, если число рекомендаций было меньше N
3. Выполнен тюнинг гиперпараметров, векторого расстояния и подбор типов моделей:
**Ноутбуки**:
HW3.1 - обоснование схемы валидации и работа с библиотекой implicit (очень долго работает на валидации, поэтому от реcерча при помощи данной библиотеки ушли), формирование взвешивания item-ов по механизму idf
HW-3.2-rectools-research - research библиотеки rectools, работает намного быстрее, чем implicit, поэтому валидацию, подбор гиперпараметров и векторного расстояния решили проводить с ней.
HW-3.3-rectools-cv - валидация и выбор оптимальной модели, формирование оффлайн рекомендаций, объяснение схемы блендинга при помощи библиотеки rectools.
HW-3.4-model-for-online-recs - обучение лучших моделей на валдиации (TFIDFRecommender(K=60), BM25Recommender(K=60)) для онлайн рекомендаций при помощи библиотеки implicit.
HW-3-test - ноутбук для тестирования нагрузки на сервис.

4. Проведены эксперименты с оффлайн валидацией.
5. Для холодных пользователей выдаются популярные рекомендации.
6. Сделан механизм блендинга.
7. Сделаны оффлайн и онлайн рекомендаций. Для оффлайн блендинг работает, модель на лидерборде "blending_tfidf_bmp25_idf_offline" дала хуже результат = 0.0658971 (на лидерборде стоит 0.08808465, неверный результат произошел лаг). Для онлайн рекомендаций модели на лидерборде помечены с приставкой implicit, блендинг работает долго для онлайн рекомендаций, поэтому на лидерборде нет результатов.


 - Формирование параметров запуска рекомендаций происходит при помощи конфига inference.cfg.yml. 
 - Загрузка всех параметров для запуска происходит в service/utils/download_artifact.py
 - Формирование рекомендаций, как онлайн, так и оффлайн происходит при помощи service/utils/run_reco.py

In48semenov committed:
ДЗ-3:
1. Реализована загрузка необходимых параметров запуска
2. Реализовано offline и online рекомендации
3. Реализован блендинг моделей
4. Проведено множество экспериментов с подбором оптимального количества парметров, векторного пространства и библиотек.
5. Всё встроено в сервис.

In48semenov committed:
Убрана неиспользуемая ячейка

dimaruso commented on service/api/views.py:
```diff
@@ -71,8 +74,43 @@ async def get_reco(
+                user_id=user_id,
+                model=artifact["model_artifact_1"],
+                watched=artifact["watched"],
+                user_mapping=artifact["users_mapping"],
```
Разумно придерживаться одного нейминга в разрастающемся проекте. Если вы что-то начали называть user_mapping - плохо вводить переменные с тем же смыслом и называть их чуть-чуть по-другому (в вашем примере users_mapping). Так становится сложнее искать затрагиваемые части через поиск по проекту.


Конкретно в этом случае можно попробовать назвать параметры идентично там где этом можно и разделить загрузку моделей, но об этом в других комментариях.

In48semenov replied:
Принято, поправлю

dimaruso commented on service/utils/download_artifact.py:
```diff
@@ -0,0 +1,68 @@
+import yaml
+
+
+def get_artifact() -> Dict:
```
Загрузка всего и сразу выглядит довольно неудобно. Лучше делать разные загрузчики для разных моделей. В идеале с привязкой к самому классу модели, но это уже может быть избыточно в данном случае т.к. кучи классов пока особо нет.

In48semenov replied:
Я думал, сделать класс и там определить разные методы, но мне показалось пока смысла в этом нет. Попробую сначала вынести загрузку моделей в другие функции

dimaruso replied:
DownloadArtifact все еще выглядит довольно страшно. 

По хорошему надо разделить загрузку данных по юзерам, загрузку моделей перенести в инит каждой модели. Пути для каждой модели хранить в конфиге.

dimaruso commented on service/utils/run_reco.py:
```diff
@@ -0,0 +1,181 @@
+        elif len(recs) < k_recs:
+            recs = add_reco_popular(k_recs, recs, popular_items)
+    else:
+        recs = list(popular_items[:k_recs])
```
Все три проверки выглядят избыточно и будут дублироваться многократно. При этом нужно всегда иметь гарантии, что выдача будет содержать ровно K элементов.

Уже хорошо, что догадались вынести это в отдельную функцию add_reco_popular, но лучше ее вызов перенести в саму ручку. Замечу, что если ее вызывать всегда - то она будет работать и для полупустой выдачи и для холодных юзеров

In48semenov replied:
Хорошо, исправлю

dimaruso replied:
Вообще это замечание было про добавление в будущем отдельного пайплайна постобработки рекомендаций. 
У нас пока только одно бизнес-правило - вернуть не менее К рекомендаций. Если какая-то модель не может вернуть реко (например юзер оказался холодным) или возвращает меньше, чем К - то возвращается заглушка (популярное). Но в общем случае их много и лучше стараться все бизнес-правила вынести в один класс и вызывать отдельно после получения самих рекомендаций.

dimaruso commented on service/utils/run_reco.py:
```diff
@@ -0,0 +1,181 @@
+        if this is not enough.
+    """
+    for item_pop in popular_items:
+        if item_pop not in curr_recs:
```
Кажется лучше сделать сет и делать проверку: not in set

In48semenov replied:
Исправлю

dimaruso replied:
Ну так еще хуже стало. Теперь конструктор set вызывается каждый раз в цикле заново

dimaruso commented on service/utils/run_reco.py:
```diff
@@ -0,0 +1,181 @@
+        The function creates online recommendation.
+    """
+    if user_id in user_mapping:
+        recs = recs_mapper(
```
Если выпилить пандас и заменить его нампаем - то будет будет быстрее) 
Пандасом удобно прочитать табличку при старте сервиса и на этом все. При ниференсе его лучше не использовать

In48semenov replied:
online reco без блендинга ок работает, прям супер, даже быстрее, чем с оффлайн рекомендациями, а вот онлайн с блендингом не успевает выплевывать рекомендации. Заменю нампаем)

dimaruso commented on service/utils/run_reco.py:
```diff
@@ -0,0 +1,181 @@
+            item_idf, left_on='item_id', right_on='index', how='left'
+        ).sort_values(['idf'], ascending=False)['item_id'].tolist()
+
+        if len(recs) > k_recs:
```
Вынести дозаполнение

In48semenov replied:
принято

dimaruso commented on service/utils/user_knn.py:
```diff
@@ -0,0 +1,134 @@
+        self.use_weight_idf = use_weight_idf
+        self.is_fitted = False
+
+    def get_mappings(self, train):
```
Тогда уж set_mappings - метод ничего не возвращает. Да и если задаете переменные класса не в __init__ - то их все равно там нужно указать и добавить тип Optional[ваш тип]

dimaruso replied:
Этот метод на самом деле приватный

dimaruso commented on service/utils/user_knn.py:
```diff
@@ -0,0 +1,134 @@
+        self.items_inv_mapping = dict(enumerate(train['item_id'].unique()))
+        self.items_mapping = {v: k for k, v in self.items_inv_mapping.items()}
+
+    def get_matrix(
```
Этот метод на самом деле приватный

dimaruso commented on service/utils/user_knn.py:
```diff
@@ -0,0 +1,134 @@
+        self.watched = df.groupby(user_col).agg({item_col: list})
+        return interaction_matrix
+
+    def idf(self, n: int, x: float):
```
Этот метод на самом деле приватный

dimaruso commented on service/utils/user_knn.py:
```diff
@@ -0,0 +1,134 @@
+        del item_idf['doc_freq']
+        self.item_idf = item_idf
+
+    def fit(self, train: pd.DataFrame):
```
Обучение не должно выполняться при старте сервиса иначе вы его полчаса поднимать будете если он случайно упадет. 

Обучение в ноутбуке или в аирфлоу.

dimaruso commented on service/utils/user_knn.py:
```diff
@@ -0,0 +1,134 @@
+        if not self.is_fitted:
+            raise ValueError("Please call fit before predict")
+
+        mapper = self._generate_recs_mapper(
```
Пандас для юпитер ноутбуков. В онлайне реко только для одного юзера строятся. Много табличек крутить не надо.

dimaruso commented on notebooks/HW-3-test.ipynb:
```diff
@@ -0,0 +1,295 @@
+{
```
Тестируете не асинхронно - запросы поступают строго после того как выполнится предыдущий. Такие результаты оценки не вполне соответствуют тому как будет на самом деле.

In48semenov replied:
Принято

dimaruso reviewed (commented):
В целом с ресерч частью все хорошо, параметры и разные модельки перебрали - ту все ок. По сервису есть замечания. Править их может быть сложно, но стоит этим заняться т.к. проще будет делать последующие дз

In48semenov committed:
1. Изменена загрузка параметров запуска: переписано на класс.
2. Дозаполение популярны и выдача популярного холодным пользователям перенесена в ручку.
3. Переписан механизм выдачи для онлайн рекомендаций с пандаса на numpy.

dimaruso commented on service/api/views.py:
```diff
@@ -76,39 +77,45 @@ async def get_reco(
+                popular_items=popular_item
             )
 
     except Exception:
```
Не вижу смысла в конструкции try except - можно выпилить без особых проблем.

dimaruso commented on service/api/views.py:
```diff
@@ -76,39 +77,45 @@ async def get_reco(
+            recs = add_reco_popular(
                 k_recs=k_recs,
+                curr_recs=recs,
+                popular_items=popular_item
```
popular_item можно скрыть - он не передается как параметр в самой ручке.

Аналогично можно сказать про параметры в get_online_blending_reco и прочих

dimaruso reviewed (approved):
По сути все сделано, но в сервисе еще есть что оптимизировать - много избыточного кода в загрузке и лишних параметров у get_online_reco и т.п. внутри ручки.

Обычно предполагается не более двух проверок одного дз, и так как остались замечания по сервису проставлю 24 балла из 25.
